src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php

```diff
@@ -354,7 +354,15 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
                             throw new RuntimeException(sprintf('Cannot create an instance of %s from serialized data because the variadic parameter %s can only accept an array.', $class, $constructorParameter->name));
                         }
 
-                        $params = array_merge($params, $data[$paramName]);
+                        $variadicParameters = array_map(
+                            function ($parameterData) use ($reflectionClass, $constructorParameter, $paramName, $context, $format) {
+                                return $this->denormalizeParameter($reflectionClass, $constructorParameter, $paramName, $parameterData, $context, $format);
+                            },
+                            $data[$paramName]
+                        );
+
+                        $params = array_merge($params, $variadicParameters);
+                        unset($data[$key]);
```

Comment:
Why this change?

Reply:
Without this changes, when using the `PropertyNormalizer`, the deserialized object is successfully instantiated, but the property is overwritten again with the original `$data[$key]` array value ignoring the type hint.

This is tested in https://github.com/symfony/symfony/pull/31438/files#diff-9d3a423164acd48c3fd313cff0be945bR146